nwbwidgets/timeseries.py

```diff
@@ -966,4 +942,485 @@ def update(
+        channel_conversion_vector = (
+            time_series.channel_conversion if hasattr(time_series, "channel_conversion") else np.ones_like(data)
+        )
+
+        data = data * channel_conversion_vector * conversion_factor_scalar + offset_scalar
+    return data.flatten()
+
+
+def trialize_time_series(
+    time_series: TimeSeries,
+    trials_table: Union[DynamicTable, pd.DataFrame],
+    data_column: Optional[int] = 0,
+    start_time_shift: float = 0,
+    duration: float = 1,
+    alignment_column: str = "start_time",
+) -> pd.DataFrame:
+
+    if isinstance(trials_table, DynamicTable):
+        trials_table_df = trials_table.to_dataframe()
+    else:
+        trials_table_df = trials_table
+
+    trials_table_df = trials_table_df.reset_index().rename(columns={"id": "trial"})
+
+    # Map timestamps to the trial interval (start_time, stop_time)
+    timestamps = get_timeseries_tt(node=time_series)
+    # Get
+    values_to_align_array = trials_table_df[alignment_column].to_numpy()
+    left_trial_bound = values_to_align_array + start_time_shift
+    left_trial_bound = left_trial_bound[
+        left_trial_bound <= timestamps.max()
+    ]  # Filter values outside of the time series to avoid extra computation
+    right_trial_bound = left_trial_bound + duration
+
+    from_index_array = np.searchsorted(timestamps, left_trial_bound)
+    to_index_array = np.searchsorted(timestamps, right_trial_bound)
+
+    data_list = []
+    timestamps_list = []
+    alignment_column_list = []
+    for trial_index, (idx_start, idx_stop) in enumerate(zip(from_index_array, to_index_array)):
+        trial_data = get_time_series_data(time_series, idx_start=idx_start, idx_stop=idx_stop, data_column=data_column)
+        data_list.append(trial_data)
+        timestamps_list.append(timestamps[idx_start:idx_stop])
+        alignment_column_list.append([trials_table_df.start_time[trial_index]] * (idx_stop - idx_start))
+
+    data_dict = {
+        "data": np.concatenate(data_list),
+        "timestamps": np.concatenate(timestamps_list),
+        f"{alignment_column}": np.concatenate(alignment_column_list),
+    }
+
+    data_df = pd.DataFrame(data_dict)
+
+    data_df_trialized = pd.merge(left=data_df, right=trials_table_df, on=alignment_column, how="left")
+    data_df_trialized["centered_timestamps"] = data_df_trialized.timestamps - data_df_trialized[alignment_column]
+
+    return data_df_trialized
+
+
+def create_empty_figure(text="No data"):
+    empty_figure = go.Figure()
+    empty_figure.update_layout(
+        xaxis={"visible": False},
+        yaxis={"visible": False},
+        annotations=[
+            {
+                "text": f"{text}",
+                "xref": "paper",
+                "yref": "paper",
+                "showarrow": False,
+                "font": {"size": 28},
+            }
+        ],
+    )
+
+    return empty_figure
+
+
+def calculate_moving_average_over_trials(df, moving_average_window):
```

Comment:
This seems specific to your widget. Maybe it would be better as a static method.

Reply:
I wanted to separate the logic of plot creation (here in the function `build_faceting_figure`) from the logic of the widget. In my view this separation of concerns (composition) makes it easier to modify and test the plotting function. Reason being:
1) It is easier for someone who does not know the function to get the plotting function and just use it by itself to understand its behavior. In opposition to this, in the `AlignMultiTraceTimeSeriesByTrialsVariable` you need to go very deep in the function to understand what is the plotting function.
2) The plotting function can be tested or modified with minimal modification to the widget class. I think keeping the points of contact between the plotting function and all the other widget class responsibilities small leads to better design. 

Basically I view the widget class a place with the following responsabilities:
1) Data loading from the appropriate neuro data type.
2) Control definition (maybe can be done with composition using your concept of foreign controlers).
3) Handling the interaction between controls and the (external) plotting function.

What do you think? Are there downsides that I am not seeing for this separation?

Reply:
That said, I can create a plotting class and make all the methods that are floating now by themselves methods of that class to give the code more coherence.

Reply:
I think a better way to organize would be to separate calculation from plotting.

Reply:
Thanks, this got me thinking:
#239